### PR TITLE
Add Dart SPIR-V transpiler

### PR DESCRIPTION
**Pull request by chriscraws**
This is the first step in implementing flutter.dev/go/shaders.

Code changes originate from https://github.com/chriscraws/engine/tree/shader-cc,
which has a working end-to-end draft implementation.

This change includes a dart library intended to be included as an
internal dependency (`dart:_spirv`) for `dart:ui`.

Dart has been chosen over C++ for the following reasons:

- Simpler to include with the HTML renderer.
- Code will be tree-shaken for applications that don't use it.
- AOT Dart performance is comparable to performance of C++
implementation.
- Easier to read.

The transpiler currently supports a minimal set of features - planned additions for later PRs are structured types, control flow, and 2d-sampling. This should be after the feature is fully integrated.

Golden shader files are tested manually, but will be tested fully
by Flutter engine unit tests as the library is included by `dart:ui`.

Most of the files included in this change are tests and test-data.

https://github.com/flutter/flutter/issues/58361
https://github.com/flutter/flutter/issues/30763

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.
- [x] The reviewer has submitted any presubmit flakes in this PR using the [engine presubmit flakes form] before re-triggering the failure.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/master/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[engine presubmit flakes form]: https://forms.gle/Wc1VyFRYJjQTH6w5A

### CODE REVIEW

**Comment by brianosman** [comment]
> skls supports float3x2 and other floatNxM where N != M, should they be listed here?

Public SkSL (the flavor that's accepted by SkRuntimeEffect) only supports square matrices (because GLSL ES 1.00 only supports square matrices).

**Comment by chinmaygarde** [comment]
cc @zanderso

**Review by chinmaygarde — COMMENTED** [review]
Initial notes on testing and benchmarking as I review the other bits in detail. These are mostly around process and integration:

Authoring new tests currently requires `spirv-opt` and `glslangValidator` on the developers paths. There are no instructions on how the developer might fetch these and which version of these tools may be necessary. However, these tools or their equivalents are already present in the source tree via the SwiftShader dependency. After adding GN rules for the appropriate targets, you can integrate this with generation of the goldens within the build and test process. This will avoid having to check in and maintain some (but not all) SPIRV goldens. These tests running from the GTest harness will also run in JIT and AOT modes automatically and integrate with the rest of the engine testing infrastructure monitored by sheriffs.
Instead of inventing a bespoke benchmarking harness, please use the engine Google Benchmarking harness. This harness is capable of running benchmarks automatically and submitting the results to dashboards that are already being monitored. Also, this harness will run in profile and release AOT modes whose results are more relevant to the end user (and also work around the “warm up” issues you mentioned in the current harness).
The user experience for SPIRV that does not adhere to the rules specified in the README are not clear. Please add tests that assert failure modes.
The exercise of depending on SPIRV as the portable container for shaders was partially derived from the fact that SKSL had no stability guarantees. Towards that end, having SKSL goldens in the repo makes us susceptible to breaking changes to the format causing manual out of band regeneration of SKSL goldens. Instead of checking in SKSL goldens, is it possible to generate SKSL from GLSL (via an intermediate SPIRV translation step) and adding a test that ensures that the SKSL is valid using the current Skia version with SKSL reflection reflection used for test assertions? @brianosman, is there a Skia API for SKSL reflection? If not, we should at least ensure that the SKSL is valid.

**Comment by chriscraws** [comment]
> * The exercise of depending on SPIRV as the portable container for shaders was partially derived from the fact that SKSL had no stability guarantees. Towards that end, having SKSL goldens in the repo makes us susceptible to breaking changes to the format causing manual out of band regeneration of SKSL goldens. Instead of checking in SKSL goldens, is it possible to generate SKSL from GLSL (via an intermediate SPIRV translation step) and adding a test that ensures that the SKSL is valid using the current Skia version with SKSL reflection reflection used for test assertions? @brianosman, is there a Skia API for SKSL reflection? If not, we should at least ensure that the SKSL is valid.

I think generating the intermediate goldens is fine - My plan here is to ensure conformance via flutter engine tests that applies each shader to a Canvas and checks to make sure each .spv file generates the correct colors, and catching any build issue on the way - so we'd have the exact build/compilation errors for each platform. The downside is that the tests are 'farther' from the code they're testing. Is that a reasonable approach?

For your other comments I agree and I'll take a pass at making the changes

**Comment by chinmaygarde** [comment]
> I think generating the intermediate goldens is fine...

I am not concerned about the SPIRV intermediates as that is stable anyway. I meant the `sksl.golden` files specifically. Changes to the format would need manual out-of-band regeneration of the goldens which makes rolls harder.

> My plan here is to ensure conformance via flutter engine tests that applies each shader to a Canvas and checks to make sure each .spv file generates the correct colors, and catching any build issue on the way.

I see. The bit about using pixel tests to assert correctness was not clear to me. I was assuming that tests would reflect SKSL properties to ensure correctness of translation. Is reflection not possible at all with SKSL (perhaps @brianosman can answer this)? Pixel tests are harder to maintain but we can manage it. I am more concerned about tests that assert updates to SKSL that don't affect pixel output, or, affect them too subtly to make the error clear from just the pixel output. Also, stuff like optimizations like dead constant/block/function elimination, folding, etc.. I am not sure how much Skia optimizes the SKSL on its own as well. In any case, its fine to have this be an open question to be answered later.

> For your other comments I agree and I'll take a pass at making the changes

Sounds good. Please reach out for references to existing harnesses you can repurpose.

**Comment by brianosman** [comment]
To address a few questions:
- SkSL doesn't offer full reflection of the program structure. Attempts to compile it will succeed or fail (with error messages). From there, you can reflect the kinds of metadata you need to use it (how many uniforms, their types, etc...), but not anything about the internals of each function.
- There are almost no breaking changes to SkSL these days, and that pace will have to drop to ~0, as more clients pick it up.
- Alternatively - and this may be what you want for other reasons - we have built a C++ API for constructing SkSL programmatically. It's structured as a C++ embedded DSL, so you wouldn't necessarily have reflection, but you'd be directly constructing the program elements (expressions, statements, etc...), so that might be moot. The DSL is used today for some of our internal cases, and Ethan is currently working on SkRuntimeEffect integration. You can see the API [here](https://source.chromium.org/chromium/chromium/src/+/master:third_party/skia/include/sksl/).
Happy to chat about this - there are definitely some open questions around how this should all fit together.

**Comment by chriscraws** [comment]
I've updated the PR with the following changes, based on offline discussions:

1. Removed golden testing 
2. Removed benchmarks (can add benchmarks in the proper place in a follow-up PR)
3. Added "exception tests" that compile commented spir-v assembly into spir-v binary using SPIR-V tools from the SwiftShader dependency.  This required creating an assembler C++ binary to build with Flutter's configuration. 
4. Removed a few things from the transpiler that are not used yet (only relevant for control flow, which is a later PR)

The following are things expected in follow-up PRs (order after #1 is flexible)

1. Making the library available within `dart:ui` as `dart:_spirv` and creating `fragment_shader.cc` to expose fragment shaders to dart. This change will include comprehensive pixel testing for all supported SPIR-V features.
2. Add support for basic control flow.
3. Add support for child shaders / samplers.
4. Extend support to the `web_ui` library.

**Comment by clocksmith** [comment]
Does this need another review?

**Comment by brianosman** [comment]
I'm OoO this week, just did a quick skim, but I'd really need to sit down and spend some more time. I'm curious about a few of the ops (composite assemble/extract, etc.), but without walking through the SPIRV spec (or SkSL's SPIRV backend), I can't say there's any problem.

**Comment by clocksmith** [comment]
OK sounds good

File: lib/spirv/lib/src/transpiler.dart [hunk 1]
```diff
@@ -0,0 +1,935 @@
...
+  void opAccessChain() {
+    final String type = resolveType(readWord());
+    final String name = resolveName(readWord());
+    final String base = resolveName(readWord());
+
+    out.write('$indent$type $name = $base');
+    final int count = nextPosition - position;
+    for (int i = 0; i < count; i++) {
+      final String index = resolveName(readWord());
+      out.write('[$index]');
```
Comment (brianosman) [inline lib/spirv/lib/src/transpiler.dart:686]: opAccessChain is also used for struct field access, not just indexing.

**Review by brianosman — COMMENTED** [review]
Looked over this a bit better. As far as I can tell, things all make sense. I did point out one spot where opAccessChain assumes array indexing, which isn't always the case (although if you're not supporting structs yet, then it won't matter).

**Comment by clocksmith** [comment]
Can the pixel tests, along with the FragmentShader class, be landed in a followup PR? It is in progress

**Comment by chriscraws** [comment]
@brianosman Thanks for the review- not supporting structs in this PR (for now) but I can add a comment to clarify

@clocksmith, yes, that is my hope for follow up PRs. 

> The following are things expected in follow-up PRs (order after #1 is flexible)
> 
> 
> 
> 1. Making the library available within `dart:ui` as `dart:_spirv` and creating `fragment_shader.cc` to expose fragment shaders to dart. This change will include comprehensive pixel testing for all supported SPIR-V features.
> 
> 2. Add support for basic control flow.
> 
> 3. Add support for child shaders / samplers.
> 
> 4. Extend support to the `web_ui` library.

File: lib/spirv/README.md [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+- The output can only be written to from the main function.
+- `gl_FragCoord` can only be read from the main function, and its z and w components
+  have no meaning.
+- Control flow is prohibited aside from function calls and `return`.
+  `if`, `while`, `for`, `switch`, etc.
+- No inputs from other shader stages.
+- Only float, float-vector types, and square float-matrix types.
+- Only square matrices are supported.
+- Only built-in functions present in GLSL ES 100 are used.
+- Debug symbols must be stripped, you can use the `spirv-opt` `--strip_debug` flag.
```
Comment (clocksmith) [inline lib/spirv/README.md:28]: nit: I think its --strip-debug

**Review by clocksmith — APPROVED** [review]

**Comment by Hixie** [comment]
That's a lot of Dart code on the engine side. Is it possible to pull it out as a package and just have the engine/dart boundary in this repo? (I didn't look very closely, maybe the whole of this Dart code is that boundary?) The reason I ask is that the more code we take out of the engine, the smaller the binary size footprint for users who aren't using this code, and the easier it is to fix bugs in the code going forward (e.g. the package can roll pretty independently).

File: lib/spirv/lib/spirv.dart [hunk 3]
```diff
@@ -0,0 +1,70 @@
+// Copyright 2021 The Flutter Authors. All rights reserved.
```
Comment (zanderso) [inline lib/spirv/lib/spirv.dart:1]: The year in the copyright header in files in this repo should be 2013

File: lib/spirv/pubspec.yaml [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+# Use of this source code is governed by a BSD-style license that can be
+# found in the LICENSE file.
+
+name: spirv
+publish_to: none
+
+environment:
+  sdk: '>=2.11.0 <3.0.0'
+dev_dependencies: 
+  test: any
```
Comment (zanderso) [inline lib/spirv/pubspec.yaml:11]: The dependence on `package:test` does not appear to be used.

File: testing/dart/spirv_exception_test.dart [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+// @dart = 2.6
+import 'dart:io';
+import 'dart:typed_data';
+
+import 'package:spirv/spirv.dart' as spirv;
+import 'package:path/path.dart' as path;
+import 'package:test/test.dart';
```
Comment (zanderso) [inline testing/dart/spirv_exception_test.dart:11]: This will need updates to account for https://github.com/flutter/engine/pull/26215

**Review by zanderso — COMMENTED** [review]
Make sure the Dart code is covered by the analyzer on CI by adding to https://github.com/flutter/engine/blob/master/ci/analyze.sh.

**Comment by zanderso** [comment]
> That's a lot of Dart code on the engine side. Is it possible to pull it out as a package and just have the engine/dart boundary in this repo? (I didn't look very closely, maybe the whole of this Dart code is that boundary?) The reason I ask is that the more code we take out of the engine, the smaller the binary size footprint for users who aren't using this code, and the easier it is to fix bugs in the code going forward (e.g. the package can roll pretty independently).

The transpiler is needed so that the public `dart:ui` API can accept SPIR-V even though the Skia API accepts SkSL. We will have to make sure that the inclusion in `dart:ui` is structured so that the transpiler will be tree-shaken away when the shader API is not used.

**Comment by Hixie** [comment]
Ah, yeah, if the internal API is SkSL that makes sense.

**Comment by zanderso** [comment]
@chriscraws It looks like something about your branch has confused GitHub. I believe there are some git commands you can run locally to recover from this state, but it might be simpler to create a new one.

**Comment by zanderso** [comment]
@christopherfujino suggested to me over chat that you can also re-create the branch locally from scratch, and then force push it over your existing feature branch on your mirror, and it will update the PR.

**Comment by zanderso** [comment]
It might also work to do a `git pull --rebase`, and then `git push origin <branch name> --force`.

**Comment by christopherfujino** [comment]
I was able to fix the diff by merging the latest master to the feature branch. Note, that when you merged your local `master` branch, it was still 5 weeks old. You'll have to update your local master branch first, or explicitly merge the remote master branch.

**Comment by clocksmith** [comment]
I had an unmerged PR out to merge into this branch, I have closed it just in case that was the problem.

**Comment by chriscraws** [comment]
Ok I think I fixed the git weirdness. I do think it was from rebasing on to an older branch, and a lot of weird stuff happened :sweat_smile: 

It should all tree shake correctly since the transpiler is pure dart (no adding to the shared lib), so it should only be included if someone makes use of the dart:ui FragmentShader class (not in this PR)

**Comment by chriscraws** [comment]
I've opened this PR to fix the remaining build failure:
https://github.com/flutter/buildroot/pull/454

File: lib/spirv/lib/src/transpiler.dart [hunk 6]
```diff
@@ -0,0 +1,940 @@
...
+    for (int i = 0; i < count; i++) {
+      out.write(resolveName(readWord()));
+      if (i < count - 1) {
+        out.write(', ');
+      }
+    }
+    out.writeln(');');
+  }
+
+  void parseGLSLInst(int id, int type) {
```
Comment (zanderso) [inline lib/spirv/lib/src/transpiler.dart:818]: For a switch statement of this size, a lookup in a constant map may be faster and/or more compact. Like:
```dart
const Map<int, String> _glslStd450OpNames = const <int, String>{
  _glslStd450Trunc: 'trunc',
  ...
};

const Map<int, String> _glslStd450OpArgc = const <int, int>{
  _glslStd450Trunc: 1,
  ...
};

...

void parseGLSLInst(int id, int type) {
  final int inst = readWord();
  final String opName = _glslStd450OpNames[inst];
  // throw if opName is null
  final int opArgc = _glslStd450OpArgc[inst];
  parseGLSLOp(id, type, opName, opArgc);
}
```

File: lib/spirv/lib/src/types.dart [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+String _typeName(_Type t, TargetLanguage target) {
+  switch (target) {
+    case TargetLanguage.sksl:
+      return _skslTypeName(t);
+    default:
+      return _glslTypeName(t);
+  }
+}
+
+String _skslTypeName(_Type t) {
```
Comment (zanderso) [inline lib/spirv/lib/src/types.dart:40]: ditto comment here and below about using a Map.

File: lib/spirv/test/exception_shaders/assemble.py [hunk 8]
```diff
@@ -0,0 +1,28 @@
+#!/bin/python
+
```
Comment (zanderso) [inline lib/spirv/test/exception_shaders/assemble.py:2]: Needs a copyright header.

**Review by zanderso — APPROVED** [review]
lgtm w/ suggested uses of constant maps instead of giant switch statements.

**Comment by clocksmith** [comment]
@chinmaygarde @zanderso @brianosman Can I merge this?

**Review by brianosman — APPROVED** [review]

**Comment by zanderso** [comment]
Still lgtm. Feel free to merge when ready.

**Pull request merged**